app/src/main/java/org/sagebionetworks/research/mpower/tracking/fragment/AddNoteFragment.java

```diff
@@ -0,0 +1,118 @@
+import org.sagebionetworks.research.mpower.tracking.view_model.logs.TrackingItemLog;
+import org.sagebionetworks.research.mpower.tracking.view_model.TrackingTaskViewModel;
+import org.sagebionetworks.research.presentation.model.interfaces.StepView;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import butterknife.BindView;
+
+/**
+ * This fragment displays the add note screen to the user and allows the user to edit/add a new note. It is an
+ * invariant in this fragment that there is already a log in the view model for the tracking item provided as an
+ * argument and the client of this fragment is responsible for ensuring this is correct.
+ * @param <ConfigType> The type of TrackingItemConfig
+ * @param <LogType> The type of TrackingItemLog
+ * @param <ViewModelType> The type of view model.
+ */
+public class AddNoteFragment
+        <ConfigType extends TrackingItemConfig, LogType extends NoteLog, ViewModelType extends TrackingTaskViewModel<ConfigType, LogType>>
+        extends TrackingFragment<ConfigType, LogType, ViewModelType> {
+    public static final String ARGUMENT_TRACKING_ITEM = "trackingItem";
+    private static final Logger LOGGER = LoggerFactory.getLogger(AddNoteFragment.class);
+
+    @BindView(R.id.note_input)
+    EditText noteInput;
+
+    @BindView(R.id.rs2_step_navigation_action_backward)
+    ActionButton backButton;
+
+    @BindView(R.id.rs2_step_navigation_action_forward)
+    ActionButton forwardButton;
+
+    protected TrackingItem trackingItem;
+
+    @Override
+    public void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        if (savedInstanceState == null) {
+            Bundle arguments = getArguments();
+            if (arguments != null) {
+                // noinspection unchecked
+                this.trackingItem = arguments.getParcelable(ARGUMENT_TRACKING_ITEM);
+            }
+        } else {
+            // noinspection unchecked
+            this.trackingItem = savedInstanceState.getParcelable(ARGUMENT_STEP_VIEW);
+        }
+    }
+
+    @Override
+    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
+        View result = super.onCreateView(inflater, container, savedInstanceState);
+        OnApplyWindowInsetsListener topListener = SystemWindowHelper.getOnApplyWindowInsetsListener(Direction.TOP);
+        ViewCompat.setOnApplyWindowInsetsListener(this.backButton, topListener);
+        LogType log = this.viewModel.getLog(this.trackingItem);
+        String previousNote = log.getNote();
+        if (previousNote != null) {
+            this.noteInput.setText(previousNote);
+        }
+
+        // The forward button writes the note result to the log in the view model and then navigates back to the parent.
+        this.forwardButton.setOnClickListener(view -> {
+            String note = noteInput.getText().toString();
+            NoteLog noteLog = log.copyWithNote(note);
```

Comment:
If you make the NoteLog implementation class be a "data class" in Kotlin, you get copy(), equals(), toString(), and hashCode() for free.

[Read more here](https://kotlinlang.org/docs/reference/data-classes.html)

Reply:
The NoteLog implementation is a AutoValue class. The problem is that here we are dealing with the interface not the implementation so I don't have access to the for free copy method without downcasting. I think this problem will resolve itself once the NoteFragment becomes reusable since it won't be dealing directly with the log anyway.